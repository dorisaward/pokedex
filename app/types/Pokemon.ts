import * as z from "zod";

const pokemon = z.looseObject({
  /**
   * The pokedex number of the Pokemon
   */
  id: z.number(),

  /**
   * The name of the Pokemon
   */
  name: z.string(),
  /**
   * The height of the Pokemon
   */
  height: z.number(),

  /**
   * The weight of the Pokemon
   */
  weight: z.number(),

  /**
   * A set of sprites used to depict this Pokemon in the game.
   */
  sprites: z.looseObject({
    /**
     * The front of the Pokemon
     */
    front_default: z.nullish(z.string()),
  }),

  /**
   * The species of the Pokemon
   */
  species: z.looseObject({
    name: z.string(),

    url: z.string(),
  }),

  /**
   * A list of the games the Pok√©mon appeared in
   */
  game_indices: z.array(
    z.looseObject({
      /**
       * The id of the game the Pokemon appeared in
       * */
      game_index: z.number(),
    }),
  ),

  /**
   * The stats of the Pokemon
   */
  stats: z.array(
    z.looseObject({
      /**
       * The specific stat of the Pokemon
       */
      stat: z.looseObject({
        name: z.string(),

        url: z.string(),
      }),
    }),
  ),
});

export type Pokemon = z.infer<typeof pokemon>;
const parseResponse = (data: unknown): Pokemon | undefined => {
  if (!data) return undefined;
  return pokemon.parse(data);
};

export default parseResponse;
