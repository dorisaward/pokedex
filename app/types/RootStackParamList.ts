export type RootStackParamList = {
  Search: undefined;
  Details: { pokemonName: string };
  Favourites: { pokemonName: string } | undefined;
};
