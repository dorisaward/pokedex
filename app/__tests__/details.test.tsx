import Details from "@/app/details";
import { RootStackParamList } from "@/app/types/RootStackParamList";
import { RouteProp } from "@react-navigation/core";
import { NativeStackNavigationProp } from "@react-navigation/native-stack";
import { render, cleanup } from "@testing-library/react-native";
import { Pokemon } from "@/app/types/Pokemon";
import { PropsWithChildren } from "react";

afterEach(cleanup);

const mockPokemon: Pokemon = {
  height: 0,
  id: 0,
  name: "Mock Pokemon",
  species: {
    name: "mock species",
    url: "",
  },
  sprites: {},
  stats: [
    {
      stat: {
        name: "mock stat",
        url: "",
      },
      effort: 0,
      base_stat: 0,
    },
  ],
  weight: 0,
};

jest.mock("@/app/components/DisplayPokemonDetails", () =>
  // eslint-disable-next-line react/display-name
  (props: { pokemonPromise: Promise<Pokemon> }) => (
    <>{JSON.stringify(props.pokemonPromise)}</>
  ),
);

jest.mock("@/app/api", () => ({
  fetchPokemon: jest.fn().mockResolvedValue(mockPokemon),
}));

jest.mock("@/app/components/ErrorBoundary", () =>
  // eslint-disable-next-line react/display-name
  (props: PropsWithChildren) => <>{props.children}</>,
);

describe("details", () => {
  it("renders", async () => {
    // Given
    const pokemonName = "mew";
    const navigation: NativeStackNavigationProp<
      RootStackParamList,
      "Details",
      undefined
    > = {} as NativeStackNavigationProp<
      RootStackParamList,
      "Details",
      undefined
    >;
    const route: RouteProp<RootStackParamList, "Details"> = {
      key: "Details",
      name: "Details",
      path: undefined,
      params: { pokemonName },
    };
    const renderable = <Details navigation={navigation} route={route} />;

    // When
    const { toJSON } = render(renderable);

    // Then
    expect(toJSON()).toMatchSnapshot();
  });
});
