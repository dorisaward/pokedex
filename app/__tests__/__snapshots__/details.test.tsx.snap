// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`details renders 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  {}
</View>
`;
